File: DynamicPrograming/go/tribonacci.go
package dynamicprogramming

// Tribonacci returns nth tribonacci number.
func Tribonacci(n int, memo map[int]int) int {
	if memo == nil {
		memo = make(map[int]int)
	}
	if n == 0 {
		return 0
	}
	if n == 1 || n == 2 {
		return 1
	}
	if val, ok := memo[n]; ok {
		return val
	}
	memo[n] = Tribonacci(n-1, memo) + Tribonacci(n-2, memo) + Tribonacci(n-3, memo)
	return memo[n]
}

File: DynamicPrograming/ts/tribonacci.ts
export function tribonacci(n: number, memo: Record<number, number> = {}): number {
  if (n === 0 || n === 1) return n;
  if (n === 2) return 1;
  if (memo[n] !== undefined) return memo[n];
  memo[n] = tribonacci(n - 1, memo) + tribonacci(n - 2, memo) + tribonacci(n - 3, memo);
  return memo[n];
}
